### PR TITLE
[Merged by Bors] - feat: `findIdx` lemmas

### PR DESCRIPTION
**Pull request by Parcly-Taxel**
Split from #9781.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/Indexes.lean [hunk 1]
```diff
@@ -253,6 +253,68 @@ theorem findIdxs_eq_map_indexesValues (p : α → Prop) [DecidablePred p] (as :
...
+theorem findIdx_lt_length {p : α → Bool} {xs : List α} :
+    xs.findIdx p < xs.length ↔ ∃ x ∈ xs, p x := by
+  rw [← not_iff_not, not_lt]
+  have := @le_antisymm_iff _ _ (xs.findIdx p) xs.length
+  simp only [findIdx_le_length, true_and] at this
+  rw [← this, findIdx_eq_length, not_exists]
+  simp only [Bool.not_eq_true, not_and]
+
+/-- If `xs.findIdx p = i` then `p` does not hold for elements with indices less than `i`. -/
+theorem not_of_lt_findIdx {p : α → Bool} {xs : List α} {i : ℕ} (h : i < xs.findIdx p) :
```
Comment (YaelDillies) [inline Mathlib/Data/List/Indexes.lean:284]: It's confusing that `i` doesn't mean the same thing in the docstring as in the statement

File: Mathlib/Data/List/Indexes.lean [hunk 2]
```diff
@@ -253,6 +253,68 @@ theorem findIdxs_eq_map_indexesValues (p : α → Prop) [DecidablePred p] (as :
...
+
+theorem le_findIdx_of_not {p : α → Bool} {xs : List α} {i : ℕ} (h : i < xs.length)
+    (h2 : ∀ (j) (hji : j < i), ¬p (xs.get ⟨j, hji.trans h⟩)) :
+    i ≤ xs.findIdx p := by
+  by_contra! f
+  exact absurd (@findIdx_get _ p xs (f.trans h)) (h2 (xs.findIdx p) f)
+
+theorem eq_findIdx_of_not {p : α → Bool} {xs : List α} {i : ℕ} (h : i < xs.length)
+    (h2 : ∀ (j) (hji : j < i), ¬p (xs.get ⟨j, hji.trans h⟩))
+    (h3 : p (xs.get ⟨i, h⟩)) : xs.findIdx p = i := by
```
Comment (YaelDillies) [inline Mathlib/Data/List/Indexes.lean:311]:
```suggestion
theorem findIdx_eq {p : α → Bool} {xs : List α} {i : ℕ} (h : i < xs.length) :
    xs.findIdx p = i ↔ p (xs.get ⟨i, h⟩) ∧ ∀ (j) (hji : j < i), ¬p (xs.get ⟨j, hji.trans h⟩) := by
```

File: Mathlib/Data/List/Indexes.lean (hunk 1 above)
Comment (Parcly-Taxel) [reply Mathlib/Data/List/Indexes.lean]: Fixed.

**Review by YaelDillies — APPROVED** [review]
maintainer merge

File: Mathlib/Data/List/Indexes.lean [hunk 3]
```diff
@@ -253,6 +253,68 @@ theorem findIdxs_eq_map_indexesValues (p : α → Prop) [DecidablePred p] (as :
...
+    cases' i.eq_zero_or_pos with e e
+    · simpa only [e, Fin.zero_eta, get_cons_zero]
+    · have ipm := Nat.succ_pred_eq_of_pos e
+      have ilt := ho.trans_le (findIdx_le_length p)
+      rw [(Fin.mk_eq_mk (h' := ipm ▸ ilt)).mpr ipm.symm, get_cons_succ]
+      rw [← ipm, Nat.succ_lt_succ_iff] at h
+      exact ih h
+
+theorem le_findIdx_of_not {p : α → Bool} {xs : List α} {i : ℕ} (h : i < xs.length)
+    (h2 : ∀ j (hji : j < i), ¬p (xs.get ⟨j, hji.trans h⟩)) : i ≤ xs.findIdx p := by
```
Comment (eric-wieser) [inline Mathlib/Data/List/Indexes.lean:304]: Is there an `lt` version of this lemma too, which has `≤` in `h2`?

Comment (Parcly-Taxel) [reply Mathlib/Data/List/Indexes.lean:304]: Done.

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**